### ISSUE TITLE
Implement Permission Templates

### ISSUE THREAD

**Issue by JoshStrobl** [flarum/core#1753]
## Feature Request

For Flarum installations with a large amount of primary and secondary tags, it can become cumbersome to manage each individual tag's permissions, especially when some of those tags may effectively be duplicate in capabilities (not duplicate in the purpose of those tags).

**Use Cases:**

Here is a screenshot of some of the tags I've already set up permissions for. This is not a complete list, as it is fairly time consuming to set up these permissions and we have a large amount of tags (read: still a WIP).

![flarum-permissions](https://user-images.githubusercontent.com/156574/53684293-2a4f6f80-3d14-11e9-8b80-53b23eac59df.jpg)

1. We may utilize secondary tags to indicate an answer to a question (should the question be the discussion topic and the answer be something we can define in a Yes or No form). These tags should only be able to be set by specific groups (for us it's Mods, Core Team, and Global Maintainers), but people should be able to engage in those posts (as shown in the permissions set in the screenshot), react, etc. Ideally this would be something we could have as a Permissions Template, we just set the tag to the template and have a singular place to update the capabilities of the permissions template.
2. Similar in function but separate in purpose, we also have secondary tags for "Over Dramatic", "Helpful", "Needs More Information", etc. Almost all of these should be set to use the same template as Yes / No. It is different in purpose because we may use this for a tutorial, or support query, where we'd want to determine whether or not certain information is necessary (or if something is helpful, to give another example), not leave that to be a capability for the community.
3. It would be helpful to use Permission Templates on our various Primary Tags, like Announcements. This permission template would be literally what I've added in the screenshot, so only certain groups could make, sticky, etc. discussions with those tags.

I suspect this would function by adding a separate section in Permissions, maybe underneath the Groups list, for permission templates. You would define permissions that are identical to the list you already have for tags. You would have a new dropdown item above the "Read" section of the Permissions -> Tags table called "Template", where you could choose:

- None (do not apply any template)
- Template Name...
- More Items...

**Justify why this feature belongs in Flarum's core, rather than in a third-party extension**

I believe this belongs as a core feature because tags and permissions are a fundamental part of Flarum, it's something third-party extensions integrate into, and therefore the best place to implement it (as a UX and API level) would be core.

**Describe alternatives you've considered**

I do not know of any alternative mechanisms that would address this.

**Comment by franzliedke** [comment]
Just so I understand fully... the *core need* this would be addressing is to reduce the work of managing the detailed permission when you have lots of tags, right? So this is a matter of scalability in a way.

I am asking because I've repeatedly see that concrete proposals at an early stage inhibit creativity when it comes to UX / features. Starting from the problem to be solved can open up more radical (and better?) alternatives that would otherwise not be considered.

**Comment by JoshStrobl** [comment]
> Just so I understand fully... the core need this would be addressing is to reduce the work of managing the detailed permission when you have lots of tags, right? So this is a matter of scalability in a way.

Yes, that is what my proposal would accomplish.

**Comment by tobscure** [comment]
Yes the scalability of the permissions grid sucks. I mocked up a new UI concept to address this a few months ago. The idea is that you configure global permissions, and then rather than re-configuring permissions for each tag like in the grid, you can just add/remove them selectively relative to the global permissions:

![Screen Shot 2019-03-24 at 12 34 19 pm](https://user-images.githubusercontent.com/128862/54873952-8bf58c00-4e31-11e9-98d5-9e5f9019b5f4.png)

Still a WIP as to exactly what the UI would look like to create these rules, but I think this is a step in the right direction.

**Comment by luceos** [comment]
Please for the new design also consider the consequence of large amount of permission groups, not just tags.

**Comment by JoshStrobl** [comment]
Bad bot.

**Comment by Ornanovitch** [comment]
Hey all!

Is it possible to consider introducing a temporary workaround before a future complete UI revamp? As discussed in #2810 it could be really helpful at least to stick both the horizontal and vertical headers, like on [this fiddle](https://jsfiddle.net/qwubvg9m/1/)
